File: ticket-bot.py
import discord
from discord.ext import commands
from discord.ui import View, Button, Select
from collections import defaultdict


intents = discord.Intents.all()
bot = commands.Bot(command_prefix="dio!", intents=intents)


ticket_counter = 0
active_tickets = {}
staff_ratings = defaultdict(list)


CATEGORY_ID = 1397234544368685269
ALS_ROLE_ID = 1397023739631108188
ASTDX_ROLE_ID = 1382679356975087647
PROMOTION_ROLE_ID = 1397185106975920138
LOG_CHANNEL_ID = 1397579436274090035
STAFF_ROLE_ID = 1397186487044411522




class RatingButton(Button):
    def __init__(self, rating, staff_id, ticket_channel, rater_id):
        super().__init__(label=f"{rating} ⭐", style=discord.ButtonStyle.secondary)
        self.rating = rating
        self.staff_id = staff_id
        self.ticket_channel = ticket_channel
        self.rater_id = rater_id


    async def callback(self, interaction: discord.Interaction):
        if self.staff_id:
            staff_ratings[self.staff_id].append(self.rating)
            await interaction.response.send_message(f"✅ You rated {self.rating} stars!", ephemeral=True)


            log_channel = bot.get_channel(LOG_CHANNEL_ID)
            if log_channel:
                rater = interaction.guild.get_member(self.rater_id)
                staff = interaction.guild.get_member(self.staff_id)
                await log_channel.send(
                    f"⭐ **Rating**: `{self.rating} stars`\n👤 From: {rater.mention}\n🎯 To: {staff.mention}"
                )


            # Promotion check
            guild = interaction.guild
            staff_member = guild.get_member(self.staff_id)
            if staff_member and self.rating == 5:
                if staff_ratings[self.staff_id].count(5) == 15:
                    role = guild.get_role(PROMOTION_ROLE_ID)
                    if role and role not in staff_member.roles:
                        await staff_member.add_roles(role)
                        await log_channel.send(f"🎉 {staff_member.mention} has been promoted with 15 five-star ratings!")


            await self.ticket_channel.delete()




class RatingView(View):
    def __init__(self, staff_id, ticket_channel, rater_id):
        super().__init__(timeout=60)
        for i in range(1, 6):
            self.add_item(RatingButton(i, staff_id, ticket_channel, rater_id))




class DeleteTicketButton(Button):
    def __init__(self, creator_id):
        super().__init__(label="🗑️ Delete Ticket", style=discord.ButtonStyle.danger)
        self.creator_id = creator_id


    async def callback(self, interaction: discord.Interaction):
        if interaction.user.id != self.creator_id:
            await interaction.response.send_message("❌ Only the ticket creator can delete this ticket.", ephemeral=True)
            return


        staff_id = active_tickets.get(interaction.channel.id, {}).get("staff_id")
        if not staff_id:
            await interaction.channel.delete()
            return


        view = RatingView(staff_id, interaction.channel, interaction.user.id)
        await interaction.response.send_message("Please rate the support you received:", view=view, ephemeral=True)




class TakeRequestButton(Button):
    def __init__(self, creator_id):
        super().__init__(label="🎯 Take Request", style=discord.ButtonStyle.primary)
        self.creator_id = creator_id


    async def callback(self, interaction: discord.Interaction):
        data = active_tickets.get(interaction.channel.id)
        if not data:
            await interaction.response.send_message("⚠️ Internal error.", ephemeral=True)
            return


        if data["staff_id"] is not None:
            await interaction.response.send_message("❌ This request has already been taken.", ephemeral=True)
            return


        if interaction.user.id == self.creator_id:
            await interaction.response.send_message("❌ You cannot take your own request.", ephemeral=True)
            return


        data["staff_id"] = interaction.user.id
        await interaction.channel.send(f"✅ This request has been taken by {interaction.user.mention}.")
        self.disabled = True
        await interaction.message.edit(view=self.view)
        await interaction.response.send_message("🎯 You claimed this request!", ephemeral=True)




class TicketTypeSelect(Select):
    def __init__(self):
        options = [
            discord.SelectOption(label="ALS", description="Open a ticket for ALS Staff"),
            discord.SelectOption(label="ASTDX", description="Open a ticket for ASTDX Staffs")
        ]
        super().__init__(placeholder="Ticket Selector", options=options)


    async def callback(self, interaction: discord.Interaction):
        global ticket_counter
        ticket_counter += 1


        ticket_type = self.values[0].lower()
        ticket_name = f"{ticket_type}-ticket-{ticket_counter:04}"
        guild = interaction.guild
        category = guild.get_channel(CATEGORY_ID)
        helpers_role = guild.get_role(STAFF_ROLE_ID)


        overwrites = {
            guild.default_role: discord.PermissionOverwrite(read_messages=False),
            interaction.user: discord.PermissionOverwrite(read_messages=True, send_messages=True, attach_files=True, embed_links=True),
            helpers_role: discord.PermissionOverwrite(read_messages=True, send_messages=True)
        }


        ticket_channel = await guild.create_text_channel(ticket_name, overwrites=overwrites, category=category)
        active_tickets[ticket_channel.id] = {
            "creator_id": interaction.user.id,
            "staff_id": None
        }


        view = View()
        view.add_item(TakeRequestButton(interaction.user.id))
        view.add_item(DeleteTicketButton(interaction.user.id))


        role_id = ALS_ROLE_ID if ticket_type == "als" else ASTDX_ROLE_ID
        staff_role = guild.get_role(role_id)


        await ticket_channel.send(f"{interaction.user.mention} has opened a **{ticket_type.upper()}** support ticket!\n{staff_role.mention if staff_role else ''}", view=view)
        await interaction.response.send_message(f"✅ Your ticket has been created: {ticket_channel.mention}", ephemeral=True)


        # Reset selection
        self.view.clear_items()
        self.view.add_item(TicketTypeSelect())
        await interaction.message.edit(view=self.view)




class TicketBoardView(View):
    def __init__(self):
        super().__init__(timeout=None)
        self.add_item(TicketTypeSelect())




@bot.command()
async def ticketboard(ctx):
    view = TicketBoardView()
    await ctx.send("🎫 Select a ticket type below:", view=view)




@bot.command()
async def staffratings(ctx):
    if not staff_ratings:
        await ctx.send("No ratings yet.")
        return


    embed = discord.Embed(title="⭐ Staff Ratings", color=discord.Color.green())
    for staff_id, ratings in staff_ratings.items():
        user = await bot.fetch_user(staff_id)
        avg = sum(ratings) / len(ratings)
        embed.add_field(name=f"{user}", value=f"{len(ratings)} ratings | Avg: {avg:.2f}⭐", inline=False)
    await ctx.send(embed=embed)

File: bots/ticketBot.ts
import discord from 'discord.js';
import { logger } from '../services/logger';

const { Client, GatewayIntentBits, EmbedBuilder, ActionRowBuilder, ButtonBuilder, ButtonStyle, StringSelectMenuBuilder, StringSelectMenuOptionBuilder, PermissionFlagsBits } = discord;

const intents = [
  GatewayIntentBits.Guilds,
  GatewayIntentBits.GuildMessages,
  GatewayIntentBits.MessageContent,
  GatewayIntentBits.GuildMembers,
];

const bot = new Client({ intents });

const CATEGORY_ID = '1397234544368685269';
const ALS_ROLE_ID = '1397023739631108188';
const ASTDX_ROLE_ID = '1382679356975087647';
const PROMOTION_ROLE_ID = '1397185106975920138';
const LOG_CHANNEL_ID = '1397579436274090035';
const STAFF_ROLE_ID = '1397186487044411522';

let ticketCounter = 0;
const activeTickets = new Map<string, any>();
const staffRatings = new Map<string, number[]>();

bot.once('ready', () => {
  logger.info(`Ticket Bot logged in as ${bot.user?.tag}`);
});

bot.on('messageCreate', async (message) => {
  if (message.author.bot) return;
  if (!message.content.startsWith('dio!')) return;

  const args = message.content.slice(4).trim().split(/ +/);
  const command = args.shift()?.toLowerCase();

  if (command === 'ticketboard') {
    const selectMenu = new StringSelectMenuBuilder()
      .setCustomId('ticket_type_select')
      .setPlaceholder('Ticket Selector')
      .addOptions(
        new StringSelectMenuOptionBuilder()
          .setLabel('ALS')
          .setDescription('Open a ticket for ALS Staff')
          .setValue('als'),
        new StringSelectMenuOptionBuilder()
          .setLabel('ASTDX')
          .setDescription('Open a ticket for ASTDX Staffs')
          .setValue('astdx')
      );

    const row = new ActionRowBuilder<StringSelectMenuBuilder>()
      .addComponents(selectMenu);

    await message.channel.send({ content: '🎫 Select a ticket type below:', components: [row] });
  }

  if (command === 'staffratings') {
    if (staffRatings.size === 0) {
      await message.reply('No ratings yet.');
      return;
    }

    const embed = new EmbedBuilder()
      .setTitle('⭐ Staff Ratings')
      .setColor(0x00ff00);

    for (const [staffId, ratings] of staffRatings) {
      try {
        const user = await bot.users.fetch(staffId);
        const avg = ratings.reduce((sum, rating) => sum + rating, 0) / ratings.length;
        embed.addFields({
          name: user.username,
          value: `${ratings.length} ratings | Avg: ${avg.toFixed(2)}⭐`,
          inline: false,
        });
      } catch (error) {
        embed.addFields({
          name: 'Unknown User',
          value: `${ratings.length} ratings | Avg: ${(ratings.reduce((sum, rating) => sum + rating, 0) / ratings.length).toFixed(2)}⭐`,
          inline: false,
        });
      }
    }

    await message.channel.send({ embeds: [embed] });
  }
});

bot.on('interactionCreate', async (interaction) => {
  if (interaction.isStringSelectMenu() && interaction.customId === 'ticket_type_select') {
    const ticketType = interaction.values[0];
    ticketCounter++;

    const ticketName = `${ticketType}-ticket-${ticketCounter.toString().padStart(4, '0')}`;
    const guild = interaction.guild;
    const category = guild?.channels.cache.get(CATEGORY_ID);
    const helpersRole = guild?.roles.cache.get(STAFF_ROLE_ID);

    try {
      const ticketChannel = await guild?.channels.create({
        name: ticketName,
        type: discord.ChannelType.GuildText,
        parent: category?.id,
        permissionOverwrites: [
          {
            id: guild.roles.everyone.id,
            deny: [PermissionFlagsBits.ViewChannel],
          },
          {
            id: interaction.user.id,
            allow: [PermissionFlagsBits.ViewChannel, PermissionFlagsBits.SendMessages, PermissionFlagsBits.AttachFiles, PermissionFlagsBits.EmbedLinks],
          },
          {
            id: helpersRole?.id || '',
            allow: [PermissionFlagsBits.ViewChannel, PermissionFlagsBits.SendMessages],
          },
        ],
      });

      activeTickets.set(ticketChannel?.id || '', {
        creatorId: interaction.user.id,
        staffId: null,
      });

      const controlRow = new ActionRowBuilder<ButtonBuilder>()
        .addComponents(
          new ButtonBuilder()
            .setCustomId(`take_request_${interaction.user.id}`)
            .setLabel('🎯 Take Request')
            .setStyle(ButtonStyle.Primary),
          new ButtonBuilder()
            .setCustomId(`delete_ticket_${interaction.user.id}`)
            .setLabel('🗑️ Delete Ticket')
            .setStyle(ButtonStyle.Danger)
        );

      const roleId = ticketType === 'als' ? ALS_ROLE_ID : ASTDX_ROLE_ID;
      const staffRole = guild?.roles.cache.get(roleId);

      await ticketChannel?.send({
        content: `${interaction.user} has opened a **${ticketType.toUpperCase()}** support ticket!\n${staffRole || ''}`,
        components: [controlRow],
      });

      await interaction.reply({ content: `✅ Your ticket has been created: ${ticketChannel}`, ephemeral: true });

    } catch (error) {
      logger.error('Error creating ticket:', error);
      await interaction.reply({ content: 'Failed to create ticket.', ephemeral: true });
    }
  }

  if (interaction.isButton()) {
    const { customId, user, guild, channel } = interaction;

    if (customId.startsWith('take_request_')) {
      const creatorId = customId.split('_')[2];
      const ticketData = activeTickets.get(channel?.id || '');

      if (!ticketData) {
        await interaction.reply({ content: '⚠️ Internal error.', ephemeral: true });
        return;
      }

      if (ticketData.staffId !== null) {
        await interaction.reply({ content: '❌ This request has already been taken.', ephemeral: true });
        return;
      }

      if (user.id === creatorId) {
        await interaction.reply({ content: '❌ You cannot take your own request.', ephemeral: true });
        return;
      }

      ticketData.staffId = user.id;
      await channel?.send(`✅ This request has been taken by ${user}.`);
      await interaction.reply({ content: '🎯 You claimed this request!', ephemeral: true });

      // Disable the button
      const newRow = new ActionRowBuilder<ButtonBuilder>()
        .addComponents(
          new ButtonBuilder()
            .setCustomId(`take_request_${creatorId}`)
            .setLabel('🎯 Take Request')
            .setStyle(ButtonStyle.Primary)
            .setDisabled(true),
          new ButtonBuilder()
            .setCustomId(`delete_ticket_${creatorId}`)
            .setLabel('🗑️ Delete Ticket')
            .setStyle(ButtonStyle.Danger)
        );

      await interaction.message.edit({ components: [newRow] });
    }

    if (customId.startsWith('delete_ticket_')) {
      const creatorId = customId.split('_')[2];
      
      if (user.id !== creatorId) {
        await interaction.reply({ content: '❌ Only the ticket creator can delete this ticket.', ephemeral: true });
        return;
      }

      const ticketData = activeTickets.get(channel?.id || '');
      if (!ticketData?.staffId) {
        await channel?.delete();
        return;
      }

      // Show rating system
      const ratingRow = new ActionRowBuilder<ButtonBuilder>();
      for (let i = 1; i <= 5; i++) {
        ratingRow.addComponents(
          new ButtonBuilder()
            .setCustomId(`rate_${i}_${ticketData.staffId}_${channel?.id}_${user.id}`)
            .setLabel(`${i} ⭐`)
            .setStyle(ButtonStyle.Secondary)
        );
      }

      await interaction.reply({ content: 'Please rate the support you received:', components: [ratingRow], ephemeral: true });
    }

    if (customId.startsWith('rate_')) {
      const parts = customId.split('_');
      const rating = parseInt(parts[1]);
      const staffId = parts[2];
      const channelId = parts[3];
      const raterId = parts[4];

      if (user.id !== raterId) return;

      // Add rating
      if (!staffRatings.has(staffId)) {
        staffRatings.set(staffId, []);
      }
      staffRatings.get(staffId)?.push(rating);

      await interaction.reply({ content: `✅ You rated ${rating} stars!`, ephemeral: true });

      // Log the rating
      const logChannel = guild?.channels.cache.get(LOG_CHANNEL_ID);
      if (logChannel && logChannel.isTextBased()) {
        const rater = guild?.members.cache.get(raterId);
        const staff = guild?.members.cache.get(staffId);
        await logChannel.send(`⭐ **Rating**: \`${rating} stars\`\n👤 From: ${rater}\n🎯 To: ${staff}`);
      }

      // Check for promotion
      if (rating === 5) {
        const fiveStarCount = staffRatings.get(staffId)?.filter(r => r === 5).length || 0;
        if (fiveStarCount === 15) {
          const staffMember = guild?.members.cache.get(staffId);
          const promotionRole = guild?.roles.cache.get(PROMOTION_ROLE_ID);
          
          if (staffMember && promotionRole && !staffMember.roles.cache.has(PROMOTION_ROLE_ID)) {
            await staffMember.roles.add(promotionRole);
            const logChannel = guild?.channels.cache.get(LOG_CHANNEL_ID);
            if (logChannel && logChannel.isTextBased()) {
              await logChannel.send(`🎉 ${staffMember} has been promoted with 15 five-star ratings!`);
            }
          }
        }
      }

      // Delete the ticket
      const ticketChannel = guild?.channels.cache.get(channelId);
      await ticketChannel?.delete();
    }
  }
});

bot.on('error', (error) => {
  logger.error('Ticket Bot error:', error);
});

bot.on('disconnect', () => {
  logger.warn('Ticket Bot disconnected');
});

const token = process.env.BOT_TOKEN || process.env.TICKET_BOT_TOKEN;
if (!token) {
  logger.error('No bot token provided for Ticket Bot');
  process.exit(1);
}

bot.login(token).catch((error) => {
  logger.error('Failed to login Ticket Bot:', error);
  process.exit(1);
});
